modules/tinymce/src/core/main/ts/delete/TableDelete.ts

```diff
@@ -242,68 +240,70 @@ const isDeleteOfLastCharPos = (fromCaption: SugarElement<HTMLTableCaptionElement
         from.isEqual(last) && to.isEqual(first))
   ).getOr(true);
 
-const emptyCaretCaption = (editor: Editor, elm: SugarElement<Node>): Optional<boolean> =>
-  emptyElement(editor, elm);
+const emptyCaretCaption = (editor: Editor, elm: SugarElement<Node>): Optional<Optional<() => void>> =>
+  Optional.some(emptyElement(editor, elm));
 
-const validateCaretCaption = (rootElm: SugarElement<Node>, fromCaption: SugarElement<HTMLTableCaptionElement>, to: CaretPosition): Optional<boolean> =>
+const validateCaretCaption = (rootElm: SugarElement<Node>, fromCaption: SugarElement<HTMLTableCaptionElement>, to: CaretPosition): Optional<Optional<() => void>> =>
   getParentCaption(rootElm, SugarElement.fromDom(to.getNode()))
-    .map((toCaption) => Compare.eq(toCaption, fromCaption) === false);
+    .map((toCaption) => Compare.eq(toCaption, fromCaption) ? Optional.none() : Optional.some(Fun.noop));
 
 const deleteCaretInsideCaption = (
   editor: Editor,
   rootElm: SugarElement<Node>,
   forward: boolean,
   fromCaption: SugarElement<HTMLTableCaptionElement>,
   from: CaretPosition
-): Optional<boolean> =>
+): Optional<() => void> =>
   CaretFinder.navigate(forward, editor.getBody(), from).bind(
     (to) => isDeleteOfLastCharPos(fromCaption, forward, from, to) ?
       emptyCaretCaption(editor, fromCaption) :
       validateCaretCaption(rootElm, fromCaption, to)
-  ).or(Optional.some(true));
+  ).getOr(Optional.some(Fun.noop));
 
-const deleteCaretCells = (editor: Editor, forward: boolean, rootElm: SugarElement<Node>, startElm: SugarElement<Node>): boolean => {
+const deleteCaretCells = (editor: Editor, forward: boolean, rootElm: SugarElement<Node>, startElm: SugarElement<Node>): Optional<() => void> => {
   const from = CaretPosition.fromRangeStart(editor.selection.getRng());
   return getParentCell(rootElm, startElm).bind(
     (fromCell) => Empty.isEmpty(fromCell) ?
       emptyElement(editor, fromCell) :
       deleteBetweenCells(editor, rootElm, forward, fromCell, from)
-  ).getOr(false);
+  );
 };
 
 const deleteCaretCaption = (
   editor: Editor,
   forward: boolean,
   rootElm: SugarElement<Node>,
   fromCaption: SugarElement<HTMLTableCaptionElement>
-): Optional<boolean> => {
+): Optional<() => void> => {
   const from = CaretPosition.fromRangeStart(editor.selection.getRng());
   return Empty.isEmpty(fromCaption) ?
     emptyElement(editor, fromCaption) :
     deleteCaretInsideCaption(editor, rootElm, forward, fromCaption, from);
 };
 
-const isNearTable = (forward: boolean, pos: CaretPosition): boolean =>
-  forward ? isBeforeTable(pos) : isAfterTable(pos);
+const isNearTable = (forward: boolean, pos: CaretPosition): Optional<() => void> => {
```

Comment:
Not sure everything needs to be optional thunks since read operations shouldn't be a problem to have happening before blocking things.

Reply:
I tried to keep the generation of `Optional.some` in the same functions that previously had `return true`, as it serves a similar purpose now.